### PR TITLE
Refactor configs

### PR DESCRIPTION
**Pull request by yannjor**
Modules now have their own configs instead of everything being in
`config.rs`.

### CODE REVIEW

File: src/memory.rs [hunk 1]
```diff
@@ -34,9 +32,18 @@ impl MemoryUnit {
...
     header: String,
+    /// Unit used for memory usage. Possible values include KiB, MiB and GiB
+    unit: MemoryUnit,
+    /// Whether to show memory usage as percentage
+    percentage: bool,
+}
+
+pub struct Memory {
     usage: String,
+    config: MemoryConfig,
```
Comment (isacsund) [inline src/memory.rs]: is it possible to skip the extra struct here and use `#[serde(skip)]` for the fields we don't want to (de)serialize?

File: src/memory.rs [hunk 2]
```diff
@@ -64,7 +71,7 @@ fn parse(contents: &str) -> HashMap<String, u32> {
         .collect()
 }
 
-pub fn get_memory_usage(config: &Config) -> String {
+pub fn get_usage(config: &MemoryConfig) -> String {
```
Comment (isacsund) [inline src/memory.rs]: i think some refactoring for this function is needed so we don't rely on any config here and instead fix the usage how we want it in the `print()` function below

File: src/config.rs [hunk 3]
```diff
@@ -1,29 +1,30 @@
...
-    pub mem_unit: MemoryUnit,
-    /// Whether to show memory usage as percentage
-    pub mem_percentage: bool,
-    /// Whether to show the full path of the shell
-    pub shell_path: bool,
+    /// Accent color in output
+    #[serde(skip)]
+    pub color: Color,
+    pub memory: Memory,
+    pub shell: Shell,
```
Comment (isacsund) [inline src/config.rs]: lets have a space between these to make it more clear the skip is only for color

File: src/config.rs [hunk 4]
```diff
@@ -1,29 +1,30 @@
...
 use serde::{Deserialize, Serialize};
 
-use crate::memory::MemoryUnit;
+use crate::memory::Memory;
+use crate::shell::Shell;
 
 const BINARY_NAME: &str = env!("CARGO_PKG_NAME");
 
 #[derive(Serialize, Deserialize)]
 #[serde(default)]
```
Comment (isacsund) [inline src/config.rs:13]: note that this default only applies to **values** inside the struct (you need a separate `#[serde(default)]` for the structs)

File: src/cpu.rs [hunk 5]
```diff
@@ -1,9 +1,8 @@
-use ansi_term::Color;
-
 use std::collections::HashMap;
 use std::fs::read_to_string;
 
 use crate::Module;
+use crate::config::Config;
```
Comment (isacsund) [inline src/cpu.rs:4]: alphabetical order :)

File: src/memory.rs [hunk 6]
```diff
@@ -16,8 +14,8 @@ const MEBIBYTE: f32 = 1024.0;
 const GIBIBYTE: f32 = 1048576.0;
```
Comment (isacsund) [inline src/memory.rs]: `MEBIBYTE * 1024`

File: src/shell.rs [hunk 7]
```diff
@@ -1,44 +1,53 @@
...
 use crate::Module;
 
+#[derive(Serialize, Deserialize)]
 pub struct Shell {
-    header: String,
+    #[serde(skip)]
     shell: Option<String>,
+    header: String,
+    /// Whether to show the full path of the shell
+    path: bool,
```
Comment (isacsund) [inline src/shell.rs]: i think this could be named `show_path`
(also spacing here)

File: src/shell.rs [hunk 8]
```diff
@@ -1,44 +1,53 @@
...
 
-fn get_shell(config: &Config) -> Option<String> {
-    let shell = match env::var("SHELL") {
-        Ok(p) => p,
+fn get_shell() -> Option<String> {
+    match env::var("SHELL") {
+        Ok(s) => Some(s),
         Err(e) => {
             eprintln!("Failed to detect shell, {}", e);
             return None;
```
Comment (isacsund) [inline src/shell.rs]: `None` should be enough (check with `cargo clippy`)

**Review by isacsund — CHANGES_REQUESTED** [review]

File: src/config.rs [hunk 9]
```diff
@@ -1,29 +1,30 @@
...
 
 impl Default for Config {
     fn default() -> Self {
         Self {
-            mem_unit: MemoryUnit::GiB,
-            mem_percentage: true,
-            shell_path: false,
+            memory: Memory::default(),
+            shell: Shell::default(),
+            color: Color::Cyan,
```
Comment (isacsund) [inline src/config.rs]: lets use the order they are defined in the struct

File: src/config.rs [hunk 10]
```diff
@@ -1,29 +1,34 @@
...
-    pub mem_unit: MemoryUnit,
-    /// Whether to show memory usage as percentage
-    pub mem_percentage: bool,
-    /// Whether to show the full path of the shell
-    pub shell_path: bool,
+    /// Accent color in output
+    #[serde(skip)]
+    pub color: Color,
+
+    #[serde(default)]
```
Comment (isacsund) [inline src/config.rs]: this should be in the struct :( (check what happens if you delete a field in your `config.toml`)

Comment (yannjor) [reply src/config.rs]: Oh 😅

**Review by isacsund — APPROVED** [review]

**Pull request merged**